python_ta/cfg/visitor.py

```diff
@@ -128,6 +129,20 @@ def visit_functiondef(self, func: nodes.FunctionDef) -> None:
         self._current_cfg.start.add_statement(func.args)
         func.cfg_block = self._current_cfg.start
 
+        if func.name != "__main__":
```

Comment:
In response to your question about the `z3` import: overall you're on the right track, though the code can be simplified. The simplest way to think about it is that *this specific condition should have an `and z3 is not None` attached*. So that if `z3` fails to import, the code inside this if branch (which is responsible for populating the `_z3_vars` attribute) won't execute, but the rest of the existing code of this visitor class will.

However, you are correct in pointing out the issues that the imports span multiple files (e.g., this file doesn't import `z3` directly, just `ExprWrapper`). There are a few different ways to handle this elegantly. One option is to use the try-except strategy but set `ExprWrapper = None` if there's an `ImportError` raised, which could be a "nested error" because `ExprWrapper.py` imports `z3`.

Another options, which is my preference though slightly more involved, is to move this logic into a `ControlFlowGraph` method. Noting that the key AST node at play here is the `Arguments` node, one can imagine a method called `ControlFlowGraph.add_arguments` that takes an `Arguments` node and (1) adds it to the starting block (currently done in line 129 above) as well as extracts `z3` variables from those arguments using `ExprWrapper`. This would be the same overall logic as you have currently here, just moved into a single file.

Reply:
Oh and as a separate comment, I'm not actually sure why the `__main__` check is necessary. This method is `visit_functiondef`, so should only be called on `FunctionDef` nodes, and not the root `Module` node (or the `if __name__ == '__main__'` block).